### ISSUE TITLE
Performance regression in large Android project with Gradle 7.0-milestone-1

### ISSUE THREAD

**Issue by runningcode** [gradle/gradle#16048]
Seeing a performance regression with Gradle 7.0 in the following scenarios using the Uber Android Buildsystem evaluation project. https://github.com/uber-common/android-build-eval

```
abiChangeWithMaxLeafLocalCache
abiChangeWithAvgLeafLocalCache
abiChangeWithMinLeafLocalCache
nonAbiChangeWithMaxLeafLocalCache
androidResourceChangeWithRootLocalCache
noOpLocalCache
```

![image](https://user-images.githubusercontent.com/332597/107045407-c8cc9100-67c5-11eb-9478-19024c47f688.png)

![image](https://user-images.githubusercontent.com/332597/107045631-04675b00-67c6-11eb-9b01-95f0a681f624.png)

### Steps to Reproduce 
Checkout https://github.com/uber-common/android-build-eval
```
gradle-profiler --benchmark --gradle-version 6.7.1 --gradle-version 7.0-milestone-1 --iterations=10 --warmups=4 --scenario-file mobile_app1/performance.scenarios --project-dir mobile_app1
```

### Your Environment
Run in an AWS Ubuntu box of some sort. Can get more details if needed.

**Comment by runningcode** [comment]
Here are the benchmarks once again on our hetzner node:
![image](https://user-images.githubusercontent.com/332597/107938482-c1f00c00-6f85-11eb-9ff3-890d4c5b4961.png)

Here is the nonAbiChangeWithMaxLeafLocalCache profile scenario run with 3 warmups 1 execution:
[nonAbiChangeWithMaxLeafLocalCache.zip](https://github.com/gradle/gradle/files/5981145/nonAbiChangeWithMaxLeafLocalCache.zip)

I think the profile shows a build speed improvement in 7.0., it might need to be re-run with 10 warm ups to show the difference.

**Comment by wolfs** [comment]
Fixed via #16178 and #16197.
